Shaders/std/conetrace.glsl

```diff
@@ -30,6 +30,100 @@ vec3 tangent(const vec3 n) {
+// }
+
+vec4 traceCone(sampler3D voxels, vec3 origin, vec3 dir, const float aperture, const float maxDist) {
+	dir = normalize(dir);
+	vec4 sampleCol;
+	float dist = 1.5 * VOXEL_SIZE * voxelgiOffset;
+	float diam = dist * aperture;
+	vec3 samplePos;
+	while (sampleCol.a < 1.0 && dist < maxDist) {
+		samplePos = dir * dist + origin;
+		float mip = max(log2(diam * voxelgiResolution.x), 0);
+		vec4 mipSample = textureLod(voxels, samplePos * 0.5 + vec3(0.5), mip);
+		sampleCol += (1 - sampleCol) * mipSample;
+		dist += max(diam / 2, VOXEL_SIZE);
+		diam = dist * aperture;
+	}
+	return sampleCol;
+}
+
+vec4 traceDiffuse(const vec3 origin, const vec3 normal, sampler3D voxels) {
+	const float angleMix = 0.5f;
+	const float aperture = 0.55785173935;
+	vec3 o1 = normalize(tangent(normal));
+	vec3 o2 = normalize(cross(o1, normal));
+	vec3 c1 = 0.5f * (o1 + o2);
+	vec3 c2 = 0.5f * (o1 - o2);
+
+	#ifdef _VoxelCones1
+	return traceCone(voxels, origin, normal, aperture, MAX_DISTANCE) * voxelgiOcc;
+	#endif
+
+	#ifdef _VoxelCones3
+	vec4 col = traceCone(voxels, origin, normal, aperture, MAX_DISTANCE);
+	col += traceCone(voxels, origin, mix(normal, -o1, angleMix), aperture, MAX_DISTANCE);
+	col += traceCone(voxels, origin, mix(normal, c2, angleMix), aperture, MAX_DISTANCE);
+	return (col / 3.0) * voxelgiOcc;
+	#endif
+
+	#ifdef _VoxelCones5
+	vec4 col = traceCone(voxels, origin, normal, aperture, MAX_DISTANCE);
+	col += traceCone(voxels, origin, mix(normal, -o1, angleMix), aperture, MAX_DISTANCE);
+	col += traceCone(voxels, origin, mix(normal, -o2, angleMix), aperture, MAX_DISTANCE);
+	col += traceCone(voxels, origin, mix(normal, c1, angleMix), aperture, MAX_DISTANCE);
+	col += traceCone(voxels, origin, mix(normal, c2, angleMix), aperture, MAX_DISTANCE);
+	return (col / 5.0) * voxelgiOcc;
+	#endif
+
+	#ifdef _VoxelCones9
+	// Normal direction
+	vec4 col = traceCone(voxels, origin, normal, aperture, MAX_DISTANCE);
+	// 4 side cones
+	col += traceCone(voxels, origin, mix(normal, o1, angleMix), aperture, MAX_DISTANCE);
+	col += traceCone(voxels, origin, mix(normal, -o1, angleMix), aperture, MAX_DISTANCE);
+	col += traceCone(voxels, origin, mix(normal, o2, angleMix), aperture, MAX_DISTANCE);
+	col += traceCone(voxels, origin, mix(normal, -o2, angleMix), aperture, MAX_DISTANCE);
+	// 4 corners
+	col += traceCone(voxels, origin, mix(normal, c1, angleMix), aperture, MAX_DISTANCE);
+	col += traceCone(voxels, origin, mix(normal, -c1, angleMix), aperture, MAX_DISTANCE);
+	col += traceCone(voxels, origin, mix(normal, c2, angleMix), aperture, MAX_DISTANCE);
+	col += traceCone(voxels, origin, mix(normal, -c2, angleMix), aperture, MAX_DISTANCE);
+	return (col / 9.0) * voxelgiOcc;
+	#endif
+
+	return vec4(0.0);
+}
+
+vec3 traceSpecular(sampler3D voxels, const vec3 pos, const vec3 normal, const vec3 viewDir, const float roughness) {
+	float specularAperture = clamp(tan((3.14159265 / 2) * roughness * 0.75), 0.0174533 * 3.0, 3.14159265);
```

Comment:
Tiny question. How did you come up with this formula? 
Does it approximate sort of microfacet brdf? 🤔

Reply:
hi, I did not change any of this. There are links referred in either make_voxels.py or this file;

Reply:
Hm, seems like we might be able to approximate GGX, to match reflections & BRDF used in Armory.
Thanks for mentioning those links!